src/fparser/two/Fortran2008.py

```diff
@@ -1341,6 +1343,227 @@ def match(string):
         return None
 
 
+class Block_Construct(BlockBase):
+    """
+    Fortran 2008 Rule 807.
+
+    block-construct is block-stmt
+                            [ specification-part ]
+                            block
+                            end-block-stmt
+
+    TODO: Should disallow COMMON, EQUIVALENCE, IMPLICIT, INTENT,
+    NAMELIST, OPTIONAL, VALUE, and statement functions (C806)
+    """
+
+    subclass_names = []
+    use_names = [
+        "Block_Stmt",
+        "Specification_Part",
+        "Execution_Part_Construct",
+        "End_Block_Stmt",
+    ]
+
+    @staticmethod
+    def match(reader):
+        return BlockBase.match(
+            Block_Stmt,
+            [Specification_Part, Execution_Part_Construct],
+            End_Block_Stmt,
+            reader,
+            match_names=True,  # C810
+            strict_match_names=True,  # C810
+        )
+
+
+class Block_Stmt(StmtBase, WORDClsBase, ScopingRegionMixin):
+    """
+    Fortran 2008 Rule 808.
+
+    block-stmt is [ block-construct-name : ] BLOCK
+
+    """
+
+    subclass_names = []
+    use_names = ["Block_Construct_Name"]
+    counter = 0
+
+    @staticmethod
+    def match(string):
+        """
+        Attempts to match the supplied text with this rule.
+
+        :param str string: the text to match.
+
+        :returns: a tuple of the matched node and instance of Counter or \
+                  None if there is no match.
+        :rtype: Tuple["BLOCK", \
+                      :py:class:`fparser.two.Fortran2008.Block_Stmt.Counter`] \
+                | NoneType
+        """
+        found = WORDClsBase.match("BLOCK", None, string)
+        if not found:
+            return None
+        block, _ = found
+        # Construct a unique name for this BLOCK (in case it isn't named). We
+        # ensure the name is not a valid Fortran name so that it can't clash
+        # with any regions named in the code.
+        scope_name = f"block:{Block_Stmt.counter}"
```

Comment:
Yes, all symbol tables are stored in a dict and indexed by a region name. Possibly we could revisit that since anything below the top-level can be reached by going down the parent-child relationship.

That said, I don't fully understand what you mean when you say we will have [block.name="realblockname, "block:3"]. In what context will we have that/a developer encounter that? If the block is named then the scoping name will be the same as the block name or am I misunderstanding?

Reply:
Sorry I was not very precise in my previous comment, my worry is that for example in the src/fparser/two/tests/fortran2008/test_block.py::test_named_block if I print the generated named block I get:

`Block_Construct(Block_Stmt('BLOCK', 'block:5'), ..., End_Block_Stmt('BLOCK', Name('foo')))`

with the "block:5" and Name(foo) strings both existing somewhere in the bock subtree.

And if I do just the block_stmt I get:
```
(Pdb) block.children[0]
Block_Stmt('BLOCK', 'block:5')
(Pdb) block.children[0].get_scoped_name()
'foo'
```

Can the string in Block_Stmt be given (updated) to the proper name? or should we rely on users always using the method and not looking at the subtree itself.

Reply:
As discussed, this is a hard problem and is common to other Block constructs such as Do. I've therefore created #397 for it and added an appropriate comment to the code.